<launch>
    <arg name="num_slots" default="2" doc="" />
    <arg name="slot_duration_seconds" default="6" doc="" />
    <arg name="tdma_type" default="tdma_advanced" doc="type of TDMA node to use: tdma_advanced, or tdma_scripted or tdma_slotted_aloha" />
    <arg name="test_cycling_rates" default="true" doc="" />
    <arg name="use_hw_modem" default="true" doc="" />

    <env name="ROSCONSOLE_FORMAT" value="[${severity}] [${time:%Y-%m-%dT%H:%M:%S}] [${node}]: ${message}"/>

    <param name="use_sim_time" value="false"/>

    <group ns="modem0">
        <group if="$(arg use_hw_modem)">
            <node name="acomms_driver" pkg="ros_acomms" type="acomms_driver_node.py" respawn="true" respawn_delay="10" >
                <param name="modem_serial_port" value="/dev/ttyUSB2" type="str" />
                <param name="modem_baud_rate" value="19200" />
                <param name="set_modem_time" value="true"/>
                <param name="pwramp_txlevel" value="2" type="int" />
                <rosparam param="modem_config" subst_value="True">
                    ALL: 0
                    BND: 3
                    FC0: 25000
                    BW0: 5000
                    SST: 1
                    DTP: 50
                    pwramp.txlevel: 2
                    psk.packet.mod_hdr_version: 1
                    SRC: 0
                </rosparam>
            </node>
        </group>

        <group unless="$(arg use_hw_modem)">
            <param name="random_seed" value="1" />

            <node name="modem_sim_node" pkg="ros_acomms_modeling" type="modem_sim_node.py" output="screen" >
                <param name="center_frequency_hz" value="10000" type="int" />
                <param name="bandwidth_hz" value="5000" type="int" />
                <param name="SRC" value="0" type="int" />
            </node>

            <node name="modem_location_sim" pkg="ros_acomms_modeling" type="modem_location_sim_node.py" >
                <param name="simple_path" value="true" />
            </node>

        </group>

        <node name="acoustic_range_node" pkg="ros_acomms" type="acoustic_range_node.py" output="screen" >
            <param name="sound_speed_max_age_sec" value="30" type="int" />
        </node>

        <node name="mac_switcher" pkg="ros_acomms" type="mac_switcher_node.py" output="screen" required="true" respawn="false" clear_params="true" >
            <rosparam subst_value="True">
                default_mac_namespace: 'tdma_fast_ping'
                managed_mac_namespaces:
                    - 'tdma'

                # TODO: require managed mac nodes to continually publish the select message at minimum every managed_mac_heartbeat_timeout_sec
                # .. of the mac manager will revert to the default (or prior mac if configured that way)
                # .. mac manager select topic name: mac_switcher/MANAGED_MAC_NAMESPACE/select
                # ..              select topic type: std_msgs/Bool (True for select, False for deselect)
                managed_mac_heartbeat_timeout_sec: -1 # if set to -1, it will not timeout
            </rosparam>
        </node>

        <node name="tdma_fast_ping" pkg="ros_acomms" type="$(arg tdma_type)_node.py" output="screen" required="true" respawn="false">
            <param name="num_slots" value="$(arg num_slots)"/>
            <param name="slot_duration_seconds" value="$(arg slot_duration_seconds)" />
            <param name="active_slots" value="0,1"/>
            <param name="comms_slots" value=""/>
            <param name="pings_per_slot" value="1"/>
            <param name="ping_modem_src" value="1"/>
            <param name="ping_modem_timeout_sec" value="4" />
            <param name="packet_length_seconds" value="0"/>
            <param name="guard_time_seconds" value="0"/>
            <param name="always_send_test_data" value="true"/>
            <param name="publish_private_status" value="true"/>
            <!-- <param name="software_mute" value="true"/> -->
        </node>

        <node name="tdma" pkg="ros_acomms" type="$(arg tdma_type)_node.py" output="screen" required="true" respawn="false">
            <param name="num_slots" value="$(arg num_slots)"/>
            <param name="slot_duration_seconds" value="$(arg slot_duration_seconds)" />
            <param name="active_slots" value="0,1"/>
            <param name="packet_length_seconds" value="3"/>
            <param name="guard_time_seconds" value="2.5"/>
            <param name="publish_private_status" value="true"/>
            <param name="software_mute" value="true"/>
        </node>

        <node name="message_queue_node" pkg="ros_acomms" type="message_queue_node.py" output="screen" >
            <param name="packet_codec_file" value="$(find ros_acomms_tests)/launch/message_codec_config.yaml"/>
        </node>

        <node name="packet_dispatch" pkg="ros_acomms" type="packet_dispatch_node.py" output="screen" >
            <param name="packet_codec_file" value="$(find ros_acomms_tests)/launch/message_codec_config.yaml"/>
        </node>
    </group>

    <group ns="modem1">
        <group if="$(arg use_hw_modem)">
            <node name="acomms_driver" pkg="ros_acomms" type="acomms_driver_node.py" respawn="true" respawn_delay="10" >
                <param name="modem_serial_port" value="/dev/ttyUSB3" type="str" />
                <param name="modem_baud_rate" value="19200" />
                <param name="set_modem_time" value="true"/>
                <param name="pwramp_txlevel" value="2" type="int" />
                <rosparam param="modem_config" subst_value="True">
                    ALL: 0
                    BND: 3
                    FC0: 25000
                    BW0: 5000
                    SST: 1
                    DTP: 50
                    pwramp.txlevel: 2
                    psk.packet.mod_hdr_version: 1
                    SRC: 1
                </rosparam>
            </node>
        </group>

        <group unless="$(arg use_hw_modem)">
            <param name="random_seed" value="2" />

            <node name="modem_sim_node" pkg="ros_acomms_modeling" type="modem_sim_node.py" output="screen" >
                <param name="center_frequency_hz" value="10000" type="int" />
                <param name="bandwidth_hz" value="5000" type="int" />
                <param name="SRC" value="1" type="int" />
            </node>

            <node name="modem_location_sim" pkg="ros_acomms_modeling" type="modem_location_sim_node.py" >
                <param name="simple_path" value="true" />
            </node>

        </group>

        <node name="acoustic_range_node" pkg="ros_acomms" type="acoustic_range_node.py" output="screen" >
            <param name="sound_speed_max_age_sec" value="30" type="int" />
        </node>

        <node name="tdma" pkg="ros_acomms" type="$(arg tdma_type)_node.py" output="screen" required="true" respawn="false">
            <param name="software_mute" value="true"/>
            <param name="num_slots" value="$(arg num_slots)"/>
            <param name="slot_duration_seconds" value="$(arg slot_duration_seconds)" />
            <param name="active_slots" value="0,1"/>
            <param name="comms_slots" value=""/>
            <param name="pings_per_slot" value="1"/>
            <param name="ping_modem_src" value="0"/>
            <param name="always_send_test_data" value="True"/>
            <param name="packet_length_seconds" value="2.5"/>
            <param name="guard_time_seconds" value="2.5"/>
        </node>

        <node name="message_queue_node" pkg="ros_acomms" type="message_queue_node.py" output="screen" >
            <param name="packet_codec_file" value="$(find ros_acomms_tests)/launch/message_codec_config.yaml"/>
        </node>

        <node name="packet_dispatch" pkg="ros_acomms" type="packet_dispatch_node.py" output="screen" >
            <param name="packet_codec_file" value="$(find ros_acomms_tests)/launch/message_codec_config.yaml"/>
        </node>
    </group>

    <param name="random_seed" value="4" />
    <param name="multiplier" value="1" />
    <node name="sim_packet_performance_node" pkg="ros_acomms_modeling" type="sim_packet_performance_node.py" />

    <node name="sim_transmission_loss_node" pkg="ros_acomms_modeling" type="sim_transmission_loss_node.py"  >
        <param name="water_depth" value="20" type="int" />
        <param name="bellhop_arrivals" value="false" type="bool" />
    </node>

    <node name="clock_generator" pkg="ros_acomms_modeling" type="clock_generator.py"/>

</launch>
